### PR TITLE
Add tests for has_kernel_bundle for zero kernels and zero devices tests

### PR DESCRIPTION
**Pull request by vasilytric**
This PR provides these tests for `sycl::has_kernel_bundle`:
- Tests files with name `has_kernel_bundle_zero_devices` provides verifications for case when provide to the `sycl::has_kernel_bundle` empty vector with sycl::device
- Tests files with name `has_kernel_bundle_zero_kernels` provides verifications for case when provide to the `sycl::has_kernel_bundle` empty vector with sycl::kernel_id or application has no user-defined kernels

This PR also provides common code for kernel bundle tests and some other generic functions
___________________________________________

The `SYCL CTS CI / compile-cts (dpcpp)` failed due to `add_independent_cts_tests` CMake function is not defined in this PR, it can be found [here](https://github.com/KhronosGroup/SYCL-CTS/pull/218/files#diff-4461c617ceae0c7e0206622aacdf555aedd62eb129c2efb74c84fa1567bcbe0dR197-R214)

### CODE REVIEW

File: tests/kernel_bundle/has_kernel_bundle_zero_devices.h [hunk 1]
```diff
@@ -0,0 +1,110 @@
+/*******************************************************************************
+//
+//  SYCL 2020 Conformance Test Suite
+//
+//  Provides common code for zero devices test case
```
Comment (keryell) [inline tests/kernel_bundle/has_kernel_bundle_zero_devices.h:5]: I am puzzled by the plural of devices everywhere when there is 0 device.
Even the files names are about "devices"...

File: tests/kernel_bundle/kernel_bundle.h [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    res_type result = kernel_functor::INIT_VAL;
+    {
+      sycl::buffer<res_type, 1> res_buffer(&result, sycl::range<1>(1));
+      queue.submit([&](sycl::handler &cgh) {
+        auto acc =
+            res_buffer.template get_access<sycl::access_mode::read_write>(cgh);
+        cgh.parallel_for(sycl::range<1>(1), kernel_functor{acc});
+      });
+    }
+    return (result == kernel_functor::EXPECTED_VAL);
```
Comment (keryell) [inline tests/kernel_bundle/kernel_bundle.h:87]:
```suggestion
    return result == kernel_functor::EXPECTED_VAL;
```

File: tests/kernel_bundle/kernel_bundle.h [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+  const std::string_view m_functionName;
+  const std::string_view m_functionOverload;
+  bool m_print_bundle_state{true};
+
+ public:
+  constexpr TestCaseDescriptionBase(std::string_view functionName,
+                                    std::string_view functionOverload)
+      : m_functionName(functionName), m_functionOverload(functionOverload) {}
+
+  virtual std::string to_string() const override {
```
Comment (keryell) [inline tests/kernel_bundle/kernel_bundle.h:105]:
```suggestion
  std::string to_string() const override {
```
you are puzzling me :-)

File: tests/kernel_bundle/kernels.h [hunk 4]
```diff
@@ -0,0 +1,506 @@
...
+using namespace sycl_cts;
+
+struct kernel_base {
+  // We are using unsigned long long to make it universal for all cases
+  // including atomic64
+  using element_type = unsigned long long;
+  using accessor_t = sycl::accessor<element_type, 1,
+                                    sycl::access_mode::read_write,
+                                    sycl::target::global_buffer>;
+  static constexpr element_type INIT_VAL = 1;
```
Comment (keryell) [inline tests/kernel_bundle/kernels.h:26]: The problem with uppercase elements is that it might conflict with macros.
It should not be a problem since we control the CTS but might be a problem with a broken SYCL implementation. Anyway, such an implementation would not deserve passing the CTS. ;-)

File: util/aspect_set.cpp [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+  if (asp_set.empty()) return "none";
+
+  std::string result;
+  for (const auto &asp : asp_set) {
+    result += detail::map_name(asp);
+    result += delimiter;
+  }
+
+  // Remove the latest delimiter
+  result.resize(result.size() - delimiter.size());
```
Comment (keryell) [inline util/aspect_set.cpp:71]: We definitely want more C++20 ranges to avoid this kind of code :-)

File: util/aspect_set.cpp [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+  // Check we have any aspects to safely proceed with delimiters
+  if (asp_set.empty()) return "none";
+
+  std::string result;
+  for (const auto &asp : asp_set) {
+    result += detail::map_name(asp);
+    result += delimiter;
+  }
+
+  // Remove the latest delimiter
```
Comment (keryell) [inline util/aspect_set.cpp:70]:
```suggestion
  // Remove the latest delimiter since asp_set is not empty
```

File: util/cpp_compat.h [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+      it = container.erase(it);
+    } else {
+      ++it;
+    }
+  }
+  return old_size - container.size();
+}
+}  // namespace details
+
+/** @brief Provides std::erase_if form std::map available from c++20
```
Comment (keryell) [inline util/cpp_compat.h:38]: Good idea!
But what do you mean with `form`?

File: util/cpp_compat.h [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+//
+//  Provides advanced features not available in C++17
+//
+*******************************************************************************/
+
+#include <map>
+#include <unordered_set>
+
+#if __cplusplus >= 202002L
+#define COMPAT_CPP20 [[deprecated]]
```
Comment (keryell) [inline util/cpp_compat.h:13]: Perhaps this macro should have a more explicit name, bringing also less risk of conflict?

File: util/device_set.cpp [hunk 9]
```diff
@@ -0,0 +1,111 @@
+/*******************************************************************************
+//
+//  SYCL 2020 Conformance Test Suite
+//
+//  Provide generic device sets support for tests that require multiple
+//  devices to run
```
Comment (keryell) [inline util/device_set.cpp:6]: It is not clear to me why you cannot just use algorithms from https://en.cppreference.com/w/cpp/algorithm/set_difference and so on?
Sure the syntax is not fancy before C++20 ranges...

**Review by keryell — COMMENTED** [review]
Good work!

File: tests/common/get_cts_string.h [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+//
+*******************************************************************************/
+
+#ifndef __SYCLCTS_TESTS_COMMON_GET_CTS_STRING_H
+#define __SYCLCTS_TESTS_COMMON_GET_CTS_STRING_H
+
+#include "common.h"
+
+namespace sycl_cts {
+namespace get_cts_string {
```
[the remark this answers is not in the document]
Comment (vasilytric) [reply tests/common/get_cts_string.h]: I've applied this comment in this commit: 1e3af3f8cbaaf4afbc2d64582aed8861b2706a8e
Also this comment will be applied to another PRs

File: tests/kernel_bundle/has_kernel_bundle.h [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+#include "../../util/kernel_restrictions.h"
+#include "../common/common.h"
+#include "../common/type_coverage.h"
+#include "kernel_bundle.h"
+#include "kernels.h"
+
+
+namespace sycl_cts {
+namespace tests {
+namespace has_kernel_bundle {
```
[the remark this answers is not in the document]
Comment (vasilytric) [reply tests/kernel_bundle/has_kernel_bundle.h]: I've applied this comment in this commit: 1e3af3f8cbaaf4afbc2d64582aed8861b2706a8e

File: tests/kernel_bundle/has_kernel_bundle_zero_devices.h (hunk 1 above)
Comment (vasilytric) [reply tests/kernel_bundle/has_kernel_bundle_zero_devices.h]: "zero devices" was replaced with "zero device" in this commit: f27ce9f11df823b5b2a9c41db4d27fb0ac20b981

File: tests/kernel_bundle/kernel_bundle.h (hunk 2 above)
Comment (vasilytric) [reply tests/kernel_bundle/kernel_bundle.h]: I've applied this comment in this commit: 5808ace3d1281e284359c6abac7f9299f498dfde

File: tests/kernel_bundle/kernel_bundle.h (hunk 3 above)
Comment (vasilytric) [reply tests/kernel_bundle/kernel_bundle.h]: Yes, You is right :)
I've applied this comment in this commit: 5808ace3d1281e284359c6abac7f9299f498dfde

File: tests/kernel_bundle/kernels.h (hunk 4 above)
Comment (vasilytric) [reply tests/kernel_bundle/kernels.h]: The variables names was replaced into lowercase in this commit: 83e848f066993d91064182b48495c900047462dc

File: util/aspect_set.cpp (hunk 5 above)
Comment (vasilytric) [reply util/aspect_set.cpp:69]: Yes, unfortunately we are using C++17

File: util/aspect_set.cpp (hunk 6 above)
Comment (vasilytric) [reply util/aspect_set.cpp]: This comment was applied in this commit: bf40834160c8a1d6e778f94af5f38e06772b05fe

File: util/cpp_compat.h (hunk 7 above)
Comment (vasilytric) [reply util/cpp_compat.h]: `form` was replaced with `from` in this commit: bf40834160c8a1d6e778f94af5f38e06772b05fe

File: util/cpp_compat.h (hunk 8 above)
Comment (vasilytric) [reply util/cpp_compat.h]: The macro name was replaced with `SYCL_CTS_COMPAT_CPP20` in this commit: 5817d076d8a44b0999706eeb0401abdbf400fa7c

File: util/device_set.cpp (hunk 9 above)
Comment (vasilytric) [reply util/device_set.cpp:6]: I am not sure which logic can we re-written by using _C++ Algorithms library_, but now we using `std::copy` instead of `for each sycle`. This changes can be found in this commit: 4f771ecb8a39b3f3b855bdb362985e12a6611687

**Review by keryell — APPROVED** [review]

**Comment by keryell** [comment]
Thanks for the polishing!

**Review by ProGTX — APPROVED** [review]

**Comment by bader** [comment]
@vasilytric, please, resolve merge conflicts.

**Comment by vasilytric** [comment]
> @vasilytric, please, resolve merge conflicts.

@bader I am resolving merge conflicts now

**Comment by vasilytric** [comment]
`SYCL CTS CI / build-common-base-image` failed with the following error: `Error: Username and password required` it probably happened due to the fact using merge conflict resolver by GitHub GUI.
Should I upload any new commit to fix this error?

**Comment by psalz** [comment]
> `SYCL CTS CI / build-common-base-image` failed with the following error: `Error: Username and password required` it probably happened due to the fact using merge conflict resolver by GitHub GUI. Should I upload any new commit to fix this error?

I'm afraid this is affecting several PRs right now. I'm investigating in #247, but no leads yet.

**Comment by psalz** [comment]
I've identified the problem and pushed a fix in #248. Please do a back-merge once that PR has been merged to get CI up and running again!

**Comment by vasilytric** [comment]
> I've identified the problem and pushed a fix in #248. Please do a back-merge once that PR has been merged to get CI up and running again!

@psalz thanks for the information

File: util/device_set.h [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+ *                                                       sycl::aspect::fp16);
+ *                 auto devs_fp64 = device_set::filtered(devs_all,
+ *                                                       sycl::aspect::fp64);
+ *                 auto devs_core = devs_all;
+ *                 devs_core.removeDevsWith(sycl::aspect::fp16);
+ *                 devs_core.substract(devs_fp64);
+ *               }
+ */
+class device_set {
+  using StorageType = std::unordered_set<sycl::device>;
```
[the remark this answers is not in the document]
Comment (vasilytric) [reply util/device_set.h:40]: @bader  should we use some container, instead of `std::unordered_set` or this container will be supported later in the hipSYCL?

[the remark this answers is not in the document]
Comment (bader) [reply util/device_set.h:40]: I think @illuhad should be able to answer this.
On your side, you can check if other containers not requiring `std::hash` specialization can be used instead of `std::unordered_set` while hipSYCL is not fixed.

[the remark this answers is not in the document]
Comment (vasilytric) [reply util/device_set.h:40]: > you can check if other containers not requiring std::hash specialization can be used instead of std::unordered_set

@bader `std::set` can't be use, but std::vector can be used as a temporary solution and `std::unordered_set` will be hidden by a macro

But I suggest wait Aksel opinion

**Comment by fraggamuffin** [comment]
needs approval from AA

File: util/device_set.h (hunk 12 above)
[the remark this answers is not in the document]
Comment (bader) [reply util/device_set.h:40]: @vasilytric, according to Aksel https://github.com/illuhad/hipSYCL/issues/680 is going to be fixed by the end of next week.

[the remark this answers is not in the document]
Comment (vasilytric) [reply util/device_set.h:40]: @bader Thanks for the information. 
What we are going to do? Waiting for the fix or make a temporary solution by using `std::vector`?

[the remark this answers is not in the document]
Comment (bader) [reply util/device_set.h:40]: I suggest we wait for updated hipSYCL compiler.

[the remark this answers is not in the document]
Comment (vasilytric) [reply util/device_set.h:40]: @bader Yes, let's wait

[the remark this answers is not in the document]
Comment (keryell) [reply util/device_set.h:40]: See also internal https://gitlab.khronos.org/sycl/Specification/-/issues/586

**Comment by fraggamuffin** [comment]
Wait until checks resolved.

**Comment by keryell** [comment]
Any reason for not speculatively merge all this series? @illuhad

**Comment by bader** [comment]
> Any reason for not speculatively merge all this series? @illuhad

I think this will break pre-commit check for hipSYCL and might lead to unintentional regressions while the check is broken. 
@illuhad, it looks like merging https://github.com/KhronosGroup/SYCL-CTS/pull/260 might take some time, so can we make another PR with noncontroversial SHA update to bring fixes unblocking the merge of this series?

**Comment by psalz** [comment]
> can we make another PR with noncontroversial SHA update to bring fixes unblocking the merge of this series?

I've bumped the container version in #269, which includes the specialization of `std::hash`.

**Comment by bader** [comment]
> > can we make another PR with noncontroversial SHA update to bring fixes unblocking the merge of this series?
> 
> I've bumped the container version in #269, which includes the specialization of `std::hash`.

Great. Thanks a lot. I think we need to bump dpcpp to include https://github.com/intel/llvm/pull/5399 ([new tests fail](https://github.com/KhronosGroup/SYCL-CTS/runs/4934201569?check_suite_focus=true)) or add disable new tests using filtering.

**Comment by bader** [comment]
> > > can we make another PR with noncontroversial SHA update to bring fixes unblocking the merge of this series?
> > 
> > 
> > I've bumped the container version in #269, which includes the specialization of `std::hash`.
> 
> Great. Thanks a lot. I think we need to bump dpcpp to include [intel/llvm#5399](https://github.com/intel/llvm/pull/5399) ([new tests fail](https://github.com/KhronosGroup/SYCL-CTS/runs/4934201569?check_suite_focus=true)) or add disable new tests using filtering.

I've opened https://github.com/KhronosGroup/SYCL-CTS/pull/270 to bump dpcpp version. I hope there will be no failures with it.

**Comment by bader** [comment]
@vasilytric, could you update the branch to re-start CI checks, please? hipSYCL and dpcpp compiler has been updated, so hopefully we can pass the checks now.

**Comment by vasilytric** [comment]
> could you update the branch to re-start CI checks, please? hipSYCL and dpcpp compiler has been updated, so hopefully we can pass the checks now.

@bader All verifications, except `compile-cts (dpcpp, ec97c57)` has passed

**Comment by bader** [comment]
Great. Please, update all PR in the kernel_bundle test series.
